test/dynamo/test_modules.py

```diff
@@ -1164,6 +1165,48 @@ def fn(x):
             )
         )
 
+    def test_hooks_outer(self):
```

Comment:
Missing tests for:

1) Adding a hook, tracing, changing the hook, and ensuring a guard fails
2) What should export do with exporting a hook?
3) A hook that mutates state
4) A hook that graph breaks
5) A hook that invokes torch.compile

These might seem a little weird, but imo adding any new scope where we add support for user defined callables requires a lot of extra assurances.

Reply:
yea, i will think more about those cases but yea in general agreed we need to test this more thoroughly, including the guarding aspects.